service/interpreter/stateExecutionCounter.go

```diff
@@ -109,8 +109,37 @@ func (e *StateExecutionCounter) MarkStateIdExecutingIfNotYet(stateReqs []StateRe
 	}
 	e.totalCurrentlyExecutingCount += numOfNew
 
+	// Optimization: don't upsert SAs if currentSAsValues == stateIdCurrentlyExecutingCounts keys
+	if e.globalVersioner.IsAfterVersionOfExecutingStateIdMode() && needsUpdateSA {
+		sas, err := e.provider.GetSearchAttributes(e.ctx, []iwfidl.SearchAttributeKeyAndType{
+			{Key: ptr.Any(service.SearchAttributeExecutingStateIds),
+				ValueType: ptr.Any(iwfidl.KEYWORD_ARRAY)},
+		})
+		if err != nil {
+			e.provider.GetLogger(e.ctx).Error("error for GetSearchAttributes", err)
+		}
+
+		var currentSAsValues []string
+
+		currentSAs, ok := sas[service.SearchAttributeExecutingStateIds]
+		if ok {
+			currentSAsValues = currentSAs.StringArrayValue
+		} else {
+			e.provider.GetLogger(e.ctx).Error("search attribute IwfExecutingStateIds is not found", err)
```

Comment:
I feel like it's epxected to have empty/nil value to start with, so we probably just ignore this case, or just assign a empty slice so that `DeepEqual` can return `true` when comparing with another empty slice

Reply:
Done